pysindy/optimizers/trapping_sr3.py

```diff
@@ -297,48 +307,95 @@ def __init__(
             self.unbias = False
             self.constraint_order = constraint_order
 
+    def _set_Ptensors(self, r):
+        """Make the projection tensors used for the algorithm."""
+        N = int((r ** 2 + 3 * r) / 2.0)
+
+        # delta_{il}delta_{jk}
+        PL_tensor = np.zeros((r, r, r, N))
+        PL_tensor_unsym = np.zeros((r, r, r, N))
+        for i in range(r):
+            for j in range(r):
+                for k in range(r):
+                    for kk in range(N):
+                        if i == kk and j == k:
+                            PL_tensor_unsym[i, j, k, kk] = 1.0
+
+        # Now symmetrize PL
+        for i in range(r):
+            for j in range(N):
+                PL_tensor[:, :, i, j] = 0.5 * (
+                    PL_tensor_unsym[:, :, i, j] + PL_tensor_unsym[:, :, i, j].T
+                )
+
+        # if j == k, delta_{il}delta_{N-r+j,n}
+        # if j != k, delta_{il}delta_{r+j+k-1,n}
+        PQ_tensor = np.zeros((r, r, r, r, N))
+        for i in range(r):
+            for j in range(r):
+                for k in range(r):
+                    for kk in range(r):
+                        for n in range(N):
+                            if (j == k) and (n == N - r + j) and (i == kk):
+                                PQ_tensor[i, j, k, kk, n] = 1.0
+                            if (j != k) and (n == r + j + k - 1) and (i == kk):
+                                PQ_tensor[i, j, k, kk, n] = 1.0 / 2.0
+
+        return PL_tensor_unsym, PL_tensor, PQ_tensor
+
     def _bad_PL(self, PL):
+        """Check if PL tensor is properly defined"""
         tol = 1e-10
         return np.any((np.transpose(PL, [1, 0, 2, 3]) - PL) > tol)
 
     def _bad_PQ(self, PQ):
+        """Check if PQ tensor is properly defined"""
         tol = 1e-10
         return np.any((np.transpose(PQ, [0, 2, 1, 3, 4]) - PQ) > tol)
 
     def _check_P_matrix(self, r, n_features, N):
+        """Check if P tensor is properly defined"""
         # If these tensors are not passed, or incorrect shape, assume zeros
-        if self.PQ is None:
-            self.PQ = np.zeros((r, r, r, r, n_features))
-        elif (self.PQ).shape != (r, r, r, r, n_features) and (self.PQ).shape != (
+        if self.PQ_ is None:
+            self.PQ_ = np.zeros((r, r, r, r, n_features))
+        elif (self.PQ_).shape != (r, r, r, r, n_features) and (self.PQ_).shape != (
             r,
             r,
             r,
             r,
             N,
         ):
-            self.PQ = np.zeros((r, r, r, r, n_features))
-        if self.PL is None:
-            self.PL = np.zeros((r, r, r, n_features))
-        elif (self.PL).shape != (r, r, r, n_features) and (self.PL).shape != (
+            self.PQ_ = np.zeros((r, r, r, r, n_features))
```

Comment:
I think we probably want to show a warning when we do this so that the user knows we overrode their input.

Reply:
added warnings to all these if statements